client/verta/verta/dataset/_dataset.py

```diff
@@ -62,3 +69,88 @@ def _path_component_to_repr_lines(path_component_msg):
             lines.append("    SHA-256 checksum: {}".format(path_component_msg.sha256))
 
         return lines
+
+    def _set_commit_and_blob_path(self, commit, blob_path):
+        """
+        Associate this blob with a commit and path to enable downloads.
+
+        Parameters
+        ----------
+        commit : :class:`verta._repository.commit.Commit`
+            Commit this blob was gotten from.
+        blob_path : str
+            Location of this blob within its Repository.
+
+        """
+        self._commit = commit
+        self._blob_path = blob_path
+
+    def download(self, component_path, download_to_path=None, chunk_size=32*(10**3)):
+        """
+        Downloads `component_path` from this dataset if ModelDB-managed versioning was enabled.
+
+        Parameters
+        ----------
+        component_path : str
+            Original path of the file in this dataset to download.
+        download_to_path : str, optional
+            Path to download to. If not provided, the file(s) will be downloaded into a new path in
+            the current directory. If provided and the path already exists, it will be overwritten.
+        chunk_size : int, default 32 kB
+            Number of bytes to download at a time.
+
+        Returns
+        -------
+        downloaded_to_path : str
+            Path where file(s) were downloaded to. Identical to `download_to_path` if it was
+            provided as an argument.
+
+        """
+        if self._commit is None or self._blob_path is None:
+            raise RuntimeError(
+                "this dataset cannot be used for downloads;"
+                " consider using `commit.get()` to obtain a download-capable dataset"
+                " if ModelDB-managed versioning was enabled"
+            )
+
+        # backend will return error if `component_path` not found/versioned
+        url = self._commit._get_url_for_artifact(self._blob_path, component_path, "GET").url
+
+        # stream download to avoid overwhelming memory
+        response = _utils.make_request("GET", url, self._commit._conn, stream=True)
+        try:
+            _utils.raise_for_http_error(response)
+
+            print("downloading {} from ModelDB".format(component_path))
+            tempf = None  # declare first in case NamedTemporaryFile init fails
+            try:
+                # read response stream into temp file
+                with tempfile.NamedTemporaryFile('wb', delete=False) as tempf:
+                    for chunk in response.iter_content(chunk_size=chunk_size):
+                        tempf.write(chunk)
```

Comment:
@conradoverta I've just remembered: I put this download step _before_ the "avoid collisions with existing files" step because a file could have been placed at the desired path during download, so I wanted to generate the destination filepath after download to lessen the chances of an accidental overwrite.

What do you think?

Reply:
Moved collision avoidance + parent dir creation to before the download
https://github.com/VertaAI/modeldb/pull/780/commits/d08b434a926ca3fc4b54b155852813bc1165a82c